### ISSUE TITLE
MPV not heeding font size stylings for some fonts

### ISSUE THREAD

**Issue by magnilens** [mpv-player/mpv#12411]
### Important Information

Provide following Information:
- mpv version 0.36.0 / 20230910 Nightly (September 10, 2023)
- Windows Version Windows 10 build 19045.3448
- Source of the mpv binary https://www.videohelp.com/software/mpv-media-player
- I don't know which version of mpv this bug originated in. However, I remember seeing it in an mpv version in 2017 or 18.

### Reproduction steps

This can be reproduced by running mpv in default configuration and --no-config as well.
To reproduce, just play the sample video in both mpv and mpc-hc. 

### Expected behavior
There are 3 lines of text in 3 different fonts shown on screen. The subtitles should have the same relative size to each other regardless of what video player was used to view them.

### Actual behavior
The @FOT-SeuratCapie Pro B and Kozuka Gothic Pr6N B lines are bigger relative to the Ghandi sans line in mpv compared to mpc-hc. Mpc-hc is actually displaying the fonts in their intended sizes.

This bug seems to only affect certain fonts and these are the only two I can find at present. I can't tell what causes a certain font to have this problem but not others.

### Log file

[Log file](https://github.com/mpv-player/mpv/files/12642298/demo.txt)

### Sample files

[Sample video](https://github.com/mpv-player/mpv/files/12642309/demo.zip)

**Comment by llyyr** [comment]
Could you screenshot an example of what it looks like in mpc-hc? On my end they appear to have the same font size

**Comment by magnilens** [comment]
Here's what I see in mpc-hc
![blank+ mkv_snapshot_00 00 184](https://github.com/mpv-player/mpv/assets/145240028/7578f333-8419-450e-98ee-6a2be7475e47)

This is what it looks like in mpv
![mpv-shot0001](https://github.com/mpv-player/mpv/assets/145240028/7600c151-7404-4533-bdd7-cbdea037f164)

**Comment by TheOneric** [comment]
Thanks for the report. There is a known, unresolved [issue](https://github.com/libass/libass/issues/619) where the size of `@fonts` is sometimes off, which would explain the difference in the topmost line. But here the second line is also different but does not use the vertical `@` mode.

If with testing with MPC-HC, you are referring to MPC-HC’s internal subtitle renderer, could you please retest with xy-VSFilter or XySubFilter from [pinterf/xy-VSFilter](https://github.com/pinterf/xy-VSFilter/releases/tag/3.2.0.808) or [Cyberbeing/xy-VSFilter](https://github.com/Cyberbeing/xy-VSFilter/releases/tag/3.0.0.306)? MPC-HC’s ISR isn't always correct either.

Should the size of the `Kozuka` line in mpv still diverge from xy-VSFilter/XySubFilter’s result, please open an issue in [libass/libass](https://github.com/libass/libass/issues/new/choose) using the “Wrong rendering report” template. Upload the mpv screenshot, a screenshot from xy*, the sample video and mention which xy* version you used.

**Comment by magnilens** [comment]
The picture I took were based on mpc-hc's internal filter. I just tried pinterf's xy-vsfilter, same results. But I don't know if I hooked it into mpc-hc properly. From what I recall of using the original vsfilter over ten years ago, it was mpc-hc's option window, external filters, add filter, then select the vsfilter.dll, correct? I haven't installed it into the windows system yet. Do I need to?
That's what I did, and the font still looks the same as when using the internal filter, and I don't see any indicator that tells me I'm using the external filter instead of the internal one. However, it's not listed among the filters in the right-click menu when a video with subs is playing.

**Comment by TheOneric** [comment]
Yes, to use xy-VSFilter or XySubFilter in MPC you’ll need to register the DLL with the system (`install.bat`)). Iirc, you then also need to select VSFilter as the subtitle provider in `Playback->Output`.
MPC-HC’s ISR does not support fractional font sizes, while xy* and others do, so you could check the following. If you only see red, its MPC-HC ISR, if you see some white sticking out its xy-VSFilter, libass, ...:
```ini
[Script Info]
ScriptType: v4.00+
WrapStyle: 0
ScaledBorderAndShadow: yes
YCbCr Matrix: None
PlayResX: 1920
PlayResY: 1080
LayoutResX: 1920
LayoutResY: 1080

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: Default,Gandhi Sans,72,&H00FFFFFF,&H0000FFFF,&H00200000,&H7F404040,0,0,0,0,100,100,0,0,1,4,0,2,180,180,65,1

[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
Dialogue: 0,0:00:00.00,0:00:03.98,Default,,0,0,0,,{\an5\fscx900\fscy900\fs10\bord0\shad0}a
Dialogue: 1,0:00:00.00,0:00:03.98,Default,,0,0,0,,{\an5\fscx900\fscy900\bord0\shad0\fs10.9\1c&H0000FF}a
```
![grafik](https://github.com/mpv-player/mpv/assets/1668471/00bb3bdb-aeae-48c1-9063-fcdb8ca36237)

Alternatively, you could use Aegisub with a pre-bundled or added xy-VSFilter DLL. I see you already tried using arch1t3cht/Aegisub, but ran into issues. You should be able to circumvent them using a dummy video instead of the video from the mkv *(`Video->Use Dummy` and select the 1920×1080 preset)*.  
Make sure to select `csri/vsfilter_textsub` or `csri/xy-vsfilter_textsub` in `Andvanced->Video->Subtitle Provider` and that the required fonts are installed system-wide. If necessary you can extract the font files from the MKV with MkvToolnix.

*(I don't have access to a Windows machine, so I can't test myself, sorry)*

**Comment by magnilens** [comment]
I appear to have gotten xy-vsfilter working in mpc-hc. But I can't save a screenshot with the subtitles displayed.

The "a" has some white, showing that fractional font size is being used. But the test lines are still the same size as shown in my earlier mpc-hc screenshot.

**Comment by TheOneric** [comment]
Thanks for checking! This is indeed a libass issue then. Forwarded the report to https://github.com/libass/libass/issues/707

**Comment by astiob** [comment]
Cross-posting from the libass issue just for awareness: I can’t reproduce the Windows screenshot on my machine, and the Kozuka line appears the same in my MPC-HC as in my mpv.

What version of MPC-HC are you using?

And are you sure, for example, that you don’t have other fonts with the same name installed on your system that MPC-HC/VSFilter could be picking up?

[Here’s a modified version of your demo.mkv](https://github.com/mpv-player/mpv/files/12643312/debug.zip) that contains two changes:

* the Kozuka font is renamed to a random string to avoid clashes with any fonts you may have on your system;
* the attachments use the older, more widely supported MIME type `application/x-truetype-font`, in case you’re using an old MPC-HC that doesn’t understand the newer type `font/ttf` used in demo.mkv.

What does this look like for you?

**Comment by astiob** [comment]
> From what I recall of using the original vsfilter over ten years ago, it was mpc-hc's option window, external filters, add filter, then select the vsfilter.dll, correct?

For reference, this depends on the version of MPC-HC. The method you described sounds very old, although it might still work in current or relatively recent versions (I’m not sure). The most recent releases (from [clsid2/mpc-hc](https://github.com/clsid2/mpc-hc/releases)) have a dedicated “Subtitle renderer” dropdown in Options → Playback → Output where you just pick one renderer from a list. You do need to install it (`regsvr32 vsfilter.dll`) first, before MPC-HC can detect it. That said, I wasn’t able to get xy-VSFilter to work with modern MPC-HC the last few times I tried, only XySubFilter, but that’s usually enough for testing and for playback.

> However, it's not listed among the filters in the right-click menu when a video with subs is playing.

Yeah, this likely means the external one isn’t being used. This part hasn’t changed from the old days as far as I can tell.

> If you only see red, its MPC-HC ISR, if you see some white sticking out its xy-VSFilter, libass, ...:

I like how this test is thematically fitting in this case.

But I also want to shamelessly plug [my detect-renderer.mkv](https://chortos.selfip.net/~astiob/detect-renderer/), which is precisely intended for situations like this (besides being a fun toy). Oh, except right now it looks like a mess in the latest MPC-HC; I guess I really need to update it… and finally put it on GitHub, too.

**Comment by magnilens** [comment]
I am using the newest version of MPC-HC. It does recognize the newer MIME type for fonts. It was a relief when they finally implemented it in mpc-hc.

Poking around in my system reveals that there _is_ a @Kozuka Gothic Pr6N B font installed in my system. I think it's a windows system font. But it's an OTF, not a TTF like what I got and muxed into the mkv. A quick test specifically using the installed kozuka shows that it is _small_ like what is shown in my mpc-hc screenshots. 

Looking at the windows font preview for both the OTF and TTF versions shows them to be possibly identical in size. But the TTF has greater spacing between lines. 

I also muxed the OTF into an mkv, and now, even MPV is showing kozuka text as small, identical to its appearance in mpc-hc. So a possible explanation for the discrepancy in appearance is that mpc-hc is preferring installed system fonts while mpv is preferring embedded fonts in the event of identical fonts in both. This could present a problem for softsubbers, because even embedding fonts won't eliminate appearance discrepancy among media players. It might be beneficial to convince developers of other media players to prefer embedded over installed in the event of conflict. But that isn't an mpv problem.

Now, the question is, why is kozuka TTF bigger than kozuka OTF? 
I'm sure some of you would like to experiment with it, so I am attaching the kozuka OTF here.
[KozGoPr6N-Bold (OTF).zip](https://github.com/mpv-player/mpv/files/12644210/KozGoPr6N-Bold.OTF.zip)

However, this doesn't explain the discrepancy for the Seurat Capie font, because I definitely do not have that one or Ghandhi Sans installed on my system.

One last thing, what did you use to change the internal name of the font like that?

**Comment by astiob** [comment]
> So a possible explanation for the discrepancy in appearance is that mpc-hc is preferring installed system fonts while mpv is preferring embedded fonts in the event of identical fonts in both. This could present a problem for softsubbers, because even embedding fonts won't eliminate appearance discrepancy among media players.

Yes, this is in fact true, and sure, it is a problem. Always has been, with fansubbers forgetting to attach fonts to their releases or relying on faux bold without knowing that a real bold font also exists (but not on their machine).

For reference, libass isn’t (yet) fully consistent about this, either: if I’m not mistaken, on Linux or when explicitly choosing Fontconfig on another platform—try `mpv --sub-font-provider=fontconfig`—system fonts have the same priority as attached ones. I think it will still pick the attached one in this particular case, but only because it’s an exact match and happens to be earlier in the combined font list. In cases that involve faux bold etc., system fonts may be picked instead.

On the other hand, allowing system fonts has its benefits, too: when fansubbers do forget to attach fonts, you can just install them system-wide and enjoy the show. Which is also a thing that has commonly happened in the past, with fansubbers relying on it so much that they don’t bother re-releasing with the missing fonts attached.

> Now, the question is, why is kozuka TTF bigger than kozuka OTF?

They define completely different metric numbers:

TTF:
![Win Ascent: 1011. Win Descent: 285. Typo Ascent: 804. Typo Descent: -223. Typo Line Gap: 54. HHead Ascent: 1011. HHead Descent: -285. HHead Line Gap: 28. Capital Height: 775. X Height: 557.](https://github.com/mpv-player/mpv/assets/515193/91d06243-983e-4d74-9394-e9112d9b8b7d)

OTF:
![Win Ascent: 1426. Win Descent: 404. Really use Typo metrics. Typo Ascent: 880. Typo Descent: -120. Typo Line Gap: 1000. HHead Ascent: 880. HHead Descent: -120. HHead Line Gap: 1000. VHead Column Spacing: 1000. Capital Height: 763. X Height: 549.](https://github.com/mpv-player/mpv/assets/515193/9f4578b2-121c-4c6e-9f96-7552eb815719)

(The em size, which defines the scale on which all these numbers lie, is identical: 1000 in both fonts.)

_Why_ they do is another question, which can’t be answered just by looking at the font files. For what it’s worth, the OTF seems to be fuller and possibly newer than the TTF.

I can speculate that the original font probably came _only_ in the OTF format and someone converted it to TTF while subsetting it by removing all Japanese glyphs and most OpenType features. If that’s the case, then it’s possible that this someone also modified the metrics by hand. Don’t do that.

On the other hand, the metric values aren’t pretty, and in particular, the typo metrics differ, which wouldn’t be necessary if someone was doing this just for use with ASS, and the line gap differs massively. This suggests that maybe this is some kind of official variant of the font for the Western market, with the metrics adjusted to look better in absence of Japanese writing.

> However, this doesn't explain the discrepancy for the Seurat Capie font, because I definitely do not have that one or Ghandhi Sans installed on my system.

As Oneric mentioned, it is most likely explained by https://github.com/libass/libass/issues/619: this is an OTF font used in `@` mode, which libass is known to have incompatible metrics for. This will eventually be fixed.

> One last thing, what did you use to change the internal name of the font like that?

[fontTools](https://github.com/fonttools/fonttools)’ TTX tool:

```sh
fonttools ttx KozGoPr6n-B.ttf
# edit KozGoPr6n-B.ttx in a text editor to change the values in <name>
fonttools ttx -o xxgzwamdtidz3bvyng8mji3ne-io8t5.ttf KozGoPr6n-B.ttx
```

**Comment by magnilens** [comment]
> On the other hand, allowing system fonts has its benefits, too: when fansubbers do forget to attach fonts, you can just install them system-wide and enjoy the show.

Well, subbers have no control over what viewers have installed in their system. What they _can_ control is the content of their releases. But that's not an mpv problem, so we'll just leave it.

From this, I've learned not to assume a system font and an embedded font are the same just because they have the same name. Check the subs in both mpv and mpc-hc. If they differ, make a decision on which I want to use.
